Sensor State Widget

Penait1 opened the pull request:
## Summary
Added a widget for the homescreen that fetches the current sensors in home assistant and displays its state. It's a very basic first version, I would like to get this moving to be able to expand upon it in the future like optional icons and able to alter the sensor name text so it takes up less space. Only the systemSmall size is supported for now

## Screenshots
![Simulator Screenshot - iPhone 16 Pro - 2024-09-28 at 14 14 02](https://github.com/user-attachments/assets/5c27c998-4192-417b-b0fd-6507892d8d74)
![Simulator Screenshot - iPhone 16 Pro - 2024-09-28 at 14 14 28](https://github.com/user-attachments/assets/46e3b90d-3432-4006-893e-affd1831d56d)
![Simulator Screenshot - iPhone 16 Pro - 2024-09-28 at 14 07 07](https://github.com/user-attachments/assets/e21d1231-fd56-44f9-80d4-64ce9117c4b5)


## Any other notes
It's maybe a rough MR, I'm not that experienced in Swift/iOS development so I need some pointers how to proceed further and like to keep it small. Its functional but for example all the translation strings are missing. Also extending the Domain with the enum value `.sensors` required changes in the CarPlay section which is not ideal. Also I'm not sure about the name of the widget, maybe it just should be called Sensors instead of `DetailTable`.

bgoncal commented:
At a first glance it look good! How will it look like if user only choose one entity?
I'll find some time this week to test it myself

bgoncal commented:
I agree with naming it just "Sensors"

Penait1 commented:
I'll push some changes later this week to rename it to `Sensors`.

This is what it looks like with different amount of sensors
![Simulator Screenshot - iPhone 16 Pro - 2024-09-30 at 08 36 53](https://github.com/user-attachments/assets/289cf796-8985-49d8-b9d3-77f9a454af2b)

bgoncal commented:
Nice! When there is less sensors than the maximum, could you align them to the bottom and put a home assistant logo in the top center? Just so we make it look a bit nicer. The logo is in SharedAssets and called `Logo`

bgoncal commented:
![Simulator Screenshot - 16 Pro Paired - 2024-09-30 at 11 36 36](https://github.com/user-attachments/assets/f8d4ca6d-1a85-401d-b33c-72aae6004580)
Missing preview

bgoncal commented:
Can we  have a minimum top and bottom padding to avoid text reaching out to the edge? Like what the widget has now horizontally
![Simulator Screenshot - 16 Pro Paired - 2024-09-30 at 11 38 00](https://github.com/user-attachments/assets/41f463a6-a11a-4ab3-9816-4fdcc69ac1c2)

bgoncal commented on Sources/CarPlay/Templates/HAEntity+CarPlay.swift:
```diff
@@ -60,6 +60,8 @@ extension HAEntity {
             default:
                 break
             }
+        case .sensor:
```
You can add `.sensor` withint the `.none` case, since on press we dont have any action for the sensor

Penait1 replied:
Done

bgoncal commented on Sources/Extensions/Widgets/DetailsTable/WidgetDetailsTable.swift:
```diff
@@ -0,0 +1,32 @@
+        }
+        .contentMarginsDisabledIfAvailable()
+        // @todo Alter displayName
+        .configurationDisplayName("Sensors")
```
I imagine that by "Alter" you mean localization, but just commenting as a reminder, just add the string to Localizable.strings and run the project, then access it through `L10n.Something.something`

bgoncal commented on Sources/Extensions/Widgets/DetailsTable/WidgetDetailsTable.swift:
```diff
@@ -0,0 +1,32 @@
+        // @todo Alter displayName
+        .configurationDisplayName("Sensors")
+        // @todo Alter description
+        .description("Display state of sensors")
```
Same here

Penait1 replied:
Done

bgoncal commented on Sources/Shared/Notifications/NotificationCommands/NotificationsCommandManager.swift:
```diff
@@ -121,6 +121,8 @@ private struct HandlerUpdateWidgets: NotificationCommandHandler {
             DispatchQueue.main.async {
                 WidgetCenter.shared.reloadTimelines(ofKind: WidgetsKind.gauge.rawValue)
                 WidgetCenter.shared.reloadTimelines(ofKind: WidgetsKind.details.rawValue)
+                WidgetCenter.shared.reloadTimelines(ofKind: WidgetsKind.detailsTable.rawValue)
+
```
```suggestion
#if os(iOS)
 DataWidgetsUpdater.update()
#endif  
```

Penait1 replied:
Done

bgoncal commented on Sources/Extensions/Widgets/DetailsTable/WidgetDetailsTableView.swift:
```diff
@@ -0,0 +1,34 @@
+                if index < entry.sensorData.count - 1 {
+                    Divider()
+                        .background(Color.gray.opacity(0.5))
+                        .padding(.vertical, 8)
```
Perhaps adding `spacing` to VStack is enough?

Penait1 replied:
It was 👍

bgoncal commented:
The community will love this! 🥳

bgoncal commented:
Just a heads up, I will be on vacations for the next 2 weeks, so in case it's quiet here, thats the reason.

Penait1 commented:
I think I fixed most of your feedback. Enjoy your holiday and let me know how to improve it further if you are back

I renamed everything from `DetailTable` to `Sensors`, the preview should be fixed, added padding to prevent overflow at the top and if less than 3 sensors selected the values are aligned at the bottom and the logo is added. Enjoy your holiday and let me know how to improve it further if you are back
![Screenshot 2024-10-02 at 17 37 29](https://github.com/user-attachments/assets/e2055979-8a8d-4e73-b01a-e60a6eee49e7)

bgoncal commented on Sources/Shared/Widget/WidgetsKind.swift:
```diff
@@ -7,6 +7,7 @@ public enum WidgetsKind: String, CaseIterable {
     case gauge = "WidgetGauge"
     case details = "WidgetDetails"
     case scripts = "WidgetScripts"
+    case detailsTable = "WidgetDetailsTable"
```
```suggestion
    case sensors
```

bgoncal replied:
What we had before is an old naming, sensors is enough

bgoncal commented on Sources/Extensions/Widgets/Sensor/WidgetSensorsView.swift:
```diff
@@ -0,0 +1,53 @@
+    var body: some View {
+        VStack(alignment: .leading, spacing: 8) {
+            if entry.sensorData.count != WidgetSensorsView.maxItems {
+                Image(uiImage: Self.scaleLogo(logo: Asset.SharedAssets.logo.image, size: 30))
```
Instead of `UIGraphicsImageRenderer` can you use `.resizable` modifier and then set the frame for it? This is way less intensive operation, widget extension is very sensetive for memory and processing power usage

bgoncal commented on Sources/Shared/Notifications/NotificationCommands/NotificationsCommandManager.swift:
```diff
@@ -119,8 +119,9 @@ private struct HandlerUpdateWidgets: NotificationCommandHandler {
             DispatchQueue.main.async {
-                WidgetCenter.shared.reloadTimelines(ofKind: WidgetsKind.gauge.rawValue)
-                WidgetCenter.shared.reloadTimelines(ofKind: WidgetsKind.details.rawValue)
+                #if os(iOS)
```
I think I asked you to add this check right? If so, my bad, we also need to reload in case it is macOS, so this can be removed

bgoncal reviewed (changes requested):
Looking great, besides the 2 comments I added, just one final thing, can we auto select one sensor when the user adds the widget to home? Just so we dont present a blank widget like in my screenshot.

After these things are done, I am happy to merge it and publish to testflight

![CleanShot 2024-10-24 at 11 35 12@2x](https://github.com/user-attachments/assets/4d503644-a746-4e70-9357-b8d0a293f3b3)

bgoncal commented:
Keep an eye in TestFlight in the next days, I'll test it internally then prepare a beta

bgoncal commented:
I found another fix needed, sensors are not grouped by server, this can be very confusing for people with multiple servers configured and entities with same name. Check how other App entities group it

![image](https://github.com/user-attachments/assets/c8911dc3-b1fb-4bc6-bbd5-1839222d9852)

bgoncal commented:
Entities are not respecting the order they were selected in the widget configuration

Penait1 commented:
Ah let me see if I can look into the issues tomorrow. I didn't dare to press ready for review yet since the auto selecting of a default sensor does not work yet; The defaultResult() does not seem to work for array @Parameters.

bgoncal commented:
Oh man, my bad them, people usually forget about hitting "ready to review" thats I I merged.

Sounds like a plan!

ETE-Design commented:
@bgoncal When will this feature be out of test flight? Really looking forward to use it.

bgoncal commented:
Planned to the end of this month, currently under redesign

ETE-Design commented:
@bgoncal Is it out of Test Flight yet?

Penait1 commented:
@ETE-Design Yes it was added in version 2014.12
